### PR TITLE
Add a builtin ML implementation

### PR DESCRIPTION
**Pull request by ggeorgakoudis**
- Add implementation of decision tree, random forest
- Make OpenCV an optional dependency (ENABLE_OPENCV)
- Add env var (APOLLO_TUNING_MODEL) to select DecisionTree or RandomForest models
- Add python scripts to visualize with models with dot
- Update Kokkos connector

### CODE REVIEW

File: src/connectors/kokkos/kokkos-connector.cpp [hunk 1]
```diff
@@ -491,7 +492,7 @@ extern "C" void kokkosp_request_values(
   auto iter =
       tuning_regions.emplace(tuningProblem, std::make_pair("", nullptr));
   if (iter.second) {
-    std::string prefix = "dtree-step-0-rank-0-";
+    std::string prefix = "RandomForest-step-0-rank-0-";
```
Comment (DavidPoliakoff) [inline src/connectors/kokkos/kokkos-connector.cpp:495]: Lemme test these locally, I'll let you know if it works out

Comment (ggeorgakoudis) [reply src/connectors/kokkos/kokkos-connector.cpp:495]: Thanks! I was trying to add you as a reviewer but your name doesn't appear in the list. Any idea why?

**Review by DavidPoliakoff — APPROVED** [review]

File: src/Apollo.cpp [hunk 2]
```diff
@@ -544,102 +518,115 @@ Apollo::train(int step) {
...
+                                                 reg->model_params);
+
 
             if (Config::APOLLO_RETRAIN_ENABLE)
+#ifdef ENABLE_OPENCV
               reg->time_model =
                   ModelFactory::createRegressionTree(train_time_features,
                                                      train_time_responses);
+#else
+              assert(false && "Retraining requires OpenCV");
```
Comment (davidbeckingsale) [inline src/Apollo.cpp]: This assert could be compiled out, so mabye throw an exception instead?

Comment (ggeorgakoudis) [reply src/Apollo.cpp]: I agree that handling is not ideal. The ifdef-conditional assert is intentional. We do not (yet) support regression in our built-in ML library. The assertion will abort execution if the user enables re-training through the env var but has not compiled with ENABLE_OPENCV. Could we accept as a stopgap until regression falls in?

**Comment by cdwdirect** [comment]
Yes. I agree about implementing an internal DT.

**Review by davidbeckingsale — APPROVED** [review]

File: src/Apollo.cpp (hunk 2 above)
Comment (ggeorgakoudis) [reply src/Apollo.cpp]: Fixed through exceptions

**Pull request merged**